wolfcrypt/src/port/nxp/ksdk_port.c

```diff
@@ -405,108 +492,95 @@ int mp_mulmod(mp_int *a, mp_int *b, mp_int *c, mp_int *d)
-        if ((res = mp_mod(G, P, &tmp)) != MP_OKAY) {
-            return res;
-        }
-        G = &tmp;
-        szA = mp_unsigned_bin_size(G);
-    }
+#ifdef ENABLE_NXPLTC_TESTS
+    mp_int t;
+    mp_init(&t);
+    if (doLtcTest)
+        wolfcrypt_mp_exptmod(G, X, P, &t);
 #endif
-    szB = mp_unsigned_bin_size(X);
-    szC = mp_unsigned_bin_size(P);
-
-    if ((szA <= LTC_MAX_INT_BYTES) && (szB <= LTC_MAX_INT_BYTES) && 
-        (szC <= LTC_MAX_INT_BYTES))
-    {
-        mp_int t;
-
-        uint16_t sizeG, sizeX, sizeP;
-        uint8_t *ptrG = (uint8_t *)XMALLOC(LTC_MAX_INT_BYTES, NULL, DYNAMIC_TYPE_BIGINT);
-        uint8_t *ptrX = (uint8_t *)XMALLOC(LTC_MAX_INT_BYTES, NULL, DYNAMIC_TYPE_BIGINT);
-        uint8_t *ptrP = (uint8_t *)XMALLOC(LTC_MAX_INT_BYTES, NULL, DYNAMIC_TYPE_BIGINT);
 
-        /* if G is negative, add modulus to convert to positive number for LTC */
-        res = mp_init(&t);
-#ifndef WOLFSSL_SP_MATH
-        if (G->sign) {
-            if (res == MP_OKAY)
-                res = mp_add(G, P, &t);
-            if (res == MP_OKAY)
-                res = mp_copy(&t, G);
-        }
-#endif
+    szG = mp_unsigned_bin_size(G);
+    szX = mp_unsigned_bin_size(X);
+    szP = mp_unsigned_bin_size(P);
 
-        if (res == MP_OKAY && ptrG && ptrX && ptrP) {
+    if ((szG <= LTC_MAX_INT_BYTES) && 
+        (szX <= LTC_MAX_INT_BYTES) && 
+        (szP <= LTC_MAX_INT_BYTES))
+    {
+        uint16_t sizeG, sizeX, sizeP, sizeY;
+        uint8_t *ptrG, *ptrX, *ptrP, *ptrY;
+
+        ptrG = (uint8_t*)XMALLOC(LTC_MAX_INT_BYTES, NULL, DYNAMIC_TYPE_BIGINT);
+        ptrX = (uint8_t*)XMALLOC(LTC_MAX_INT_BYTES, NULL, DYNAMIC_TYPE_BIGINT);
+        ptrP = (uint8_t*)XMALLOC(LTC_MAX_INT_BYTES, NULL, DYNAMIC_TYPE_BIGINT);
+        ptrY = (uint8_t*)XMALLOC(LTC_MAX_INT_BYTES, NULL, DYNAMIC_TYPE_BIGINT);
+        if (ptrG && ptrX && ptrP && ptrY) {
             res = ltc_get_lsb_bin_from_mp_int(ptrG, G, &sizeG);
             if (res == MP_OKAY)
                 res = ltc_get_lsb_bin_from_mp_int(ptrX, X, &sizeX);
             if (res == MP_OKAY)
                 res = ltc_get_lsb_bin_from_mp_int(ptrP, P, &sizeP);
 
-            /* if number if greater that modulo, we must first reduce due to 
-               LTC requirement on modular exponentiation */
-            /* it needs number less than modulus.  */
-            /* we can take advantage of modular arithmetic rule that: A^B mod C = ( (A mod C)^B ) mod C
-               and so we do first (A mod N) : LTC does not give size requirement on A versus N,
-               and then the modular exponentiation.
-             */
-            /* if G >= P then */
-            if (res == MP_OKAY && LTC_PKHA_CompareBigNum(ptrG, sizeG, ptrP, sizeP) >= 0) {
-                res = (int)LTC_PKHA_ModRed(LTC_BASE, ptrG, sizeG, ptrP, sizeP,
+            /* if G >= P then reduce */
+            if (res == MP_OKAY && 
+                        LTC_PKHA_CompareBigNum(ptrG, sizeG, ptrP, sizeP) >= 0) {
```

Comment:
How expensive is it to call LTC_PKHA_CompareBigNum() as compared to LTC_PKHA_ModRed()?
That is, can we call LTC_PKHA_ModRed() always?